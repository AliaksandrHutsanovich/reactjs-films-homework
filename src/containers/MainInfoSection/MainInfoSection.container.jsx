import React, {
  memo,
  useContext,
  useEffect,
  useState,
  useCallback,
} from 'react';
import { useParams } from 'react-router-dom';
import { IMG_PATH_PREFIX } from '../../constants';
import MainInfoSection from '../../pages/FilmsPage/MainInfoSection';
import { useMovieDetailsRequest } from '../../hooks';
import { BackgroundContext } from '../../contexts';
import MovieDialog from '../MovieDialog';

import image from '../../assets/images/no-movie-poster.svg';

const MainInfoSectionContainer = () => {
  const [showInfo, setShowInfo] = useState(false);
  const [isOpen, setOpen] = useState(false);
  const { setBackground } = useContext(BackgroundContext);
  const { id } = useParams();
  const { item } = useMovieDetailsRequest(id);
  const handleShowInfo = useCallback(() => {
    setShowInfo(!showInfo);
  }, [showInfo]);

  useEffect(() => {
    if (item) {
      setBackground(`${IMG_PATH_PREFIX}${item.backdrop_path}`);
    }
  }, [item, setBackground]);

  const handleCloseDialog = useCallback(() => setOpen(false), []);
  const handleOpenDialog = useCallback(() => setOpen(true), []);

  if (!item) return null;

  return (
    <>
      <MovieDialog
        isOpen={isOpen}
        title={item.title}
        onCancel={handleCloseDialog}
        movieId={id}
      />
      <MainInfoSection
        title={item.title}
        genres={item.genres}
        rating={item.vote_average / 2}
        imgUrl={item.backdrop_path ? `${IMG_PATH_PREFIX}${item.backdrop_path}` : image}
        info={item.overview}
        isInfoShown={showInfo}
        onShowInfo={handleShowInfo}
        onWatchNow={handleOpenDialog}
      />
    </>
  );
};

export default memo(MainInfoSectionContainer);
