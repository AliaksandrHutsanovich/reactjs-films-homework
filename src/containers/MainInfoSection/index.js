export { default } from './MainInfoSection.container';
