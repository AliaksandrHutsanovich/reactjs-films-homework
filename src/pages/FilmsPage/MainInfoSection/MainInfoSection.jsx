import React, { memo, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Layout, Row, Col } from 'antd';
import { SCROLL_ID } from '../../../constants';
import {
  Text,
  HeadLine,
  List,
  Button,
  Divider,
} from '../../../components';
import { RatingStar } from '../../../patterns';

import styles from '../FilmsPage.scss';

const MainInfoSection = ({
  title,
  genres,
  rating,
  imgUrl,
  info,
  isInfoShown,
  onShowInfo,
  onWatchNow,
}) => (
  <>
    <div id={SCROLL_ID} />
    <Layout
      className={styles.layout}
      style={{
        background: `url(${imgUrl}) no-repeat`,
      }}
    >
      <Row>
        <Col span={12} className={styles.rightCol}>
          <HeadLine text={title} level={2} />
          <Row className={styles.infoWrapper}>
            <Col className={styles.listWrapper}>
              <List
                dataSource={genres}
                size="small"
                grid={{ column: 4 }}
                itemAutoWidth
                itemMarginLeft={15}
              />
            </Col>
            <Divider type="vertical" />
            <Text type="white" size={12}>
              1h 46m
            </Text>
          </Row>
          <RatingStar rating={rating} />
        </Col>
        <Col span={12} className={styles.leftCol}>
          {useMemo(() => (
            isInfoShown && (
              <Layout className={styles.textWrapper}>
                <Text type="with_shadow" size={12}>
                  {info}
                </Text>
              </Layout>
            )
          ), [isInfoShown, info])}
          <Row className={styles.buttonsWrapper}>
            <Button
              title="Watch Now"
              onClick={onWatchNow}
            />
            <Button
              title="View Info"
              onClick={onShowInfo}
              type="outlined"
            />
          </Row>
        </Col>
      </Row>
    </Layout>
  </>
);

MainInfoSection.propTypes = {
  title: PropTypes.string.isRequired,
  genres: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
  rating: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  imgUrl: PropTypes.string.isRequired,
  info: PropTypes.string.isRequired,
  isInfoShown: PropTypes.bool.isRequired,
  onShowInfo: PropTypes.func.isRequired,
  onWatchNow: PropTypes.func.isRequired,
};

export default memo(MainInfoSection);
