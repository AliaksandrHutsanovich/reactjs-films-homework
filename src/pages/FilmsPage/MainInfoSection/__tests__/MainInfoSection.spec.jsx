import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import MainInfoSection from '../MainInfoSection';

describe('<MainInfoSection />', () => {
  const renderer = new ShallowRenderer();

  it('should be rendered properly', () => {
    renderer.render(
      <MainInfoSection
        title="some movie"
        genres={[{ id: 1, name: 'genre 1' }, { id: 2, name: 'genre 2' }]}
        rating={4.5}
        imgUrl="/string"
        info="This is info about movie"
        isInfoShown
        onShowInfo={() => {}}
        onWatchNow={() => {}}
        typeInfoText="with_shadow"
      />,
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
