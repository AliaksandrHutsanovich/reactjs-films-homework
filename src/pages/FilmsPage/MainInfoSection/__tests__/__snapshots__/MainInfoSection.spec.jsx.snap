// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainInfoSection /> should be rendered properly 1`] = `
<React.Fragment>
  <div
    id="details"
  />
  <Layout
    className="layout"
    style={
      Object {
        "background": "url(/string) no-repeat",
      }
    }
  >
    <ForwardRef
      className="row"
    >
      <ForwardRef
        className="rightCol"
        span={12}
      >
        <UNDEFINED
          level={2}
          text="some movie"
        />
        <ForwardRef
          className="infoWrapper"
        >
          <ForwardRef
            className="listWrapper"
          >
            <UNDEFINED
              dataSource={
                Array [
                  Object {
                    "id": 1,
                    "name": "genre 1",
                  },
                  Object {
                    "id": 2,
                    "name": "genre 2",
                  },
                ]
              }
              grid={
                Object {
                  "column": 4,
                }
              }
              itemAutoWidth={true}
              itemMarginLeft={15}
              size="small"
            />
          </ForwardRef>
          <UNDEFINED
            type="vertical"
          />
          <UNDEFINED
            size={12}
            type="white"
          >
            1h 46m
          </UNDEFINED>
        </ForwardRef>
        <UNDEFINED
          rating={4.5}
        />
      </ForwardRef>
      <ForwardRef
        className="leftCol"
        span={12}
      >
        <Layout
          className="textWrapper"
        >
          <UNDEFINED
            size={12}
          >
            This is info about movie
          </UNDEFINED>
        </Layout>
        <ForwardRef
          className="buttonsWrapper"
        >
          <UNDEFINED
            onClick={[Function]}
            title="Watch Now"
          />
          <UNDEFINED
            onClick={[Function]}
            title="View Info"
            type="outlined"
          />
        </ForwardRef>
      </ForwardRef>
    </ForwardRef>
  </Layout>
</React.Fragment>
`;
